File: ion_crm_sales/ion_crm_sales/report/commission_to_revenue_ratio/commission_to_revenue_ratio.py
# Copyright (c) 2026, ard.ly and contributors
# For license information, please see license.txt

import frappe

from frappe import _


def execute(filters=None):

	filters = frappe._dict(filters or {})

	required = ["quarter", "fiscal_year", "company"]
	missing = [k for k in required if not filters.get(k)]
	if missing:
		frappe.throw(_("Missing required filter(s): {0}").format(", ".join(missing)))

	conditions = [
		"stcs.quarter = %(quarter)s",
		"stcs.fiscal_year = %(fiscal_year)s",
		"stcs.company = %(company)s",
	]

	query = f"""
		SELECT
			stcs.total_actual_sales,
			stcs.total_commission
		FROM `tabSales Target and Commission Sheet` stcs
		WHERE {" AND ".join(conditions)}
	"""

	data = frappe.db.sql(query, filters, as_dict=True)
	columns = [
		{"label": _("Total Actual Sales"), "fieldname": "total_actual_sales", "fieldtype": "Currency",
		 "width": 110},
		{"label": _("Total Commission"), "fieldname": "total_commission", "fieldtype": "Currency", "width": 160},
		{"label": _("Ratio"), "fieldname": "ratio", "fieldtype": "Percent",
		 "width": 300},
	]
	return columns, data

File: ion_crm_sales/ion_crm_sales/report/commission_to_revenue_ratio/commission_to_revenue_ratio.js
// Copyright (c) 2026, ard.ly and contributors
// For license information, please see license.txt

frappe.query_reports["Commission to Revenue Ratio"] = {
	filters: [
		{
			fieldname: 'company',
			label: 'Company',
			fieldtype: 'Link',
			options: 'Company',
			reqd: true,
			default: 'ION'
		},
		{
			fieldname: "quarter",
			label: __("Quarter"),
			fieldtype: "Select",
			reqd: 1,
			// Typical fiscal quarters. Adjust if you use custom quarters.
			options: "Q1\nQ2\nQ3\nQ4",
			// Default current quarter dynamically
			default: (function () {
				const m = parseInt(frappe.datetime.str_to_obj(frappe.datetime.get_today()).getMonth(), 10) + 1;
				if (m >= 1 && m <= 3) return "Q1";
				if (m >= 4 && m <= 6) return "Q2";
				if (m >= 7 && m <= 9) return "Q3";
				return "Q4";
			})(),
		},
		{
			fieldname: "fiscal_year",
			label: __("Fiscal Year"),
			fieldtype: "Link",
			options: "Fiscal Year",
			reqd: 1,
			// Try to default to current fiscal year if available
			default: frappe.datetime.get_today().split("-")[0]
		},
	],
};
